### PR TITLE
add and use iwalk

### PR DESCRIPTION
**Pull request by smichr**
#### Brief description of what is fixed or changed

SymPy is written in Python but applying SymPy functions to SymPy objects in Python containers is no cakewalk. But with `iwalk` -- or pywalk? -- this is made a little easier.

If this is added, no routines would have to be 'list aware', like `simplify`, the user could just use `iwalk(alist, simplify)`.

#### Release Notes

- core
    - `nfloat` can now be applied to arbitrary Python containers and SymPy objects
- utilities
    - `iwalk` added to `iterables` allows the application of a function to objects in arbitrary Python containers; an option to apply a function to the containers, too, is provided

### CODE REVIEW

File: sympy/utilities/iterables.py [hunk 1]
```diff
@@ -2584,3 +2584,53 @@ def rotations(s, dir=1):
...
+            new.append(iwalk(i, do, sanitize))
+        a = type(a)(new)
+    elif isinstance(a, dict):
+        a = sanitize(a) if sanitize else a
+        kv = [(
+            iwalk(k, do, sanitize) if dkeys else k,
+            iwalk(v, do, sanitize))
+            for k, v in a.items()]
+        if isinstance(a, defaultdict):
+            a.clear()
```
Comment (oscargus) [inline sympy/utilities/iterables.py]: Can you please add a test for this branch as well?

Comment (smichr) [reply sympy/utilities/iterables.py]: thanks, added

**Comment by smichr** [comment]
Any thoughts on name or functionality, @oscargus ?

**Comment by oscarbenjamin** [comment]
The best way to see if the functionality is right is by using it in more places.

For the name: iwalk suggests that it returns an iterator like islice etc. where is this consumes and returns concrete containers so it's not really about iterators. This function is like map but with the significant difference that it applies to recursively to deep data structures so a name could be map_deep (or something...)

**Comment by oscargus** [comment]
I may agree with @oscarbenjamin that maybe something along the line of `map` may be more clear. `map_deep` may indeed be better than `deepmap` as tab-completion will be more useful for the former.

**Comment by oscarbenjamin** [comment]
@smichr did you decide that this wasn't useful?

There are (unfortunately IMO) many places in SymPy where an object or a list, or a list of lists etc can be returned which makes this kind of utility useful.

**Pull request closed without merging**